src/ATen/nn_parse.py

```diff
@@ -1,49 +1,320 @@
+
+
+def get_return(args):
+    indices = [str(idx) for idx, arg in enumerate(args) if arg.get('output')]
+    return 'argument {}'.format(','.join(indices))
+
+
+ARGUMENT_MAPPINGS = {
+    'k': 'kernel_size',
+    'd': 'stride',
+    'pad': 'padding',
+    'p': 'padding',
+    'o': 'output_size',
+    'osize': 'output_size',
+    'dilation': 'dilation',
+    'adj': 'output_padding',
+    'a': 'output_padding',
 }
 
+DIMENSION_OFFSET = {
+    'width': -1,
+    'height': -2,
+    'W': -1,
+    'H': -2,
+    'T': -3,
+}
 
-def argument_to_declaration(arg):
-    typ = TYPE_TRANSLATIONS.get(arg.type, arg.type)
-    result = {
-        'arg': typ + ' ' + arg.name,
-    }
-    if arg.is_optional:
-        result['is_nullable'] = True
+SUBSTITUTIONS = {
+    'weights': 'weight',
+    'train': 'training',
+    'val': 'value',
+}
+
+
+def get_dimensionality(cname):
+    if 'Temporal' in cname:
+        return 1
+    elif 'Spatial' in cname:
+        return 2
+    elif 'Volumetric' in cname:
+        return 3
+    return None
+
+
+def camel_to_snake(name):
+    # from https://stackoverflow.com/questions/1175208/elegant-python-function-to-convert-camelcase-to-snake-case
+    s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
+    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()
+
+
+def get_thnn_args(thnn_function, params):
+    params_by_name = {p['name']: p for p in params}
+    dimensionality = get_dimensionality(thnn_function.name)
+
+    def arg_expr(prefix, suffix):
+        # e.g kW, kH
+        name = ARGUMENT_MAPPINGS[prefix]
+        if name not in params_by_name:
+            raise RuntimeError('missing arg "{}" in {}'.format(name, thnn_function.name))
+        param = params_by_name[name]
+        if param['type'] == 'IntList' and 'size' in param:
+            name = name + '_'
+        index = dimensionality + DIMENSION_OFFSET[suffix]
+        expr = '{}[{}]'.format(name, index)
+        return {'type': 'EXPRESSION', 'name': expr}
+
+    thnn_args = []
+    for arg in thnn_function.arguments:
+        name = arg.name
+        if name == 'state':
+            continue
+        aten_name = camel_to_snake(SUBSTITUTIONS.get(name, name))
+        if aten_name in params_by_name:
+            param = params_by_name[aten_name]
+            if arg.is_optional:
+                param['is_nullable'] = True
+            thnn_args.append(copy.deepcopy(param))
+        elif name[-1] in DIMENSION_OFFSET and name[:-1] in ARGUMENT_MAPPINGS:
+            # e.g kW, kH
+            thnn_args.append(arg_expr(name[:-1], name[-1]))
+        elif name == 'owidth' or name == 'oheight':
+            thnn_args.append(arg_expr(name[0], name[1:]))
+        elif name == 'scale':
+            thnn_args.append({'type': 'EXPRESSION', 'name': '1'})
+        else:
+            raise RuntimeError("{}: can't find binding for '{}'"
+                               .format(thnn_function.name, name))
+    return thnn_args
+
+
+def remove_unused_args(args, thnn_args):
```

Comment:
What is this doing?

Reply:
It returns the only the `args` whose names appear in the formal parameter list `thnn_args`. We start out with the full set of arguments listed in nn.yaml. `updateGradInput` or `accGradParameters` may take only a subset of those.

I'll add a comment.